Mathlib/CategoryTheory/Localization/CalculusOfFractions/Fractions.lean

```diff
@@ -0,0 +1,329 @@
+/-
+Copyright (c) 2024 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.CategoryTheory.Localization.CalculusOfFractions
+
+/-!
+# Lemmas on fractions
+
+Let `W : MorphismProperty C`, and objects `X` and `Y` in `C`. In this file,
+we introduce structures like `W.LeftFraction₂ X Y` which consists of two
+left fractions with the "same denominator" which shall be important in
+the construction of the preadditive structure on the localized category
+when `C` is preadditive and `W` has a left calculus of fractions.
+
+When `W` has a left calculus of fractions, we generalize the lemmas
+`RightFraction.exists_leftFraction` as `RightFraction₂.exists_leftFraction₂`,
+`Localization.exists_leftFraction` as `Localization.exists_leftFraction₂` and
+`Localization.exists_leftFraction₃`, and
+`LeftFraction.map_eq_iff` as `LeftFraction₂.map_eq_iff`.
+
+## Implementation note
+
+The lemmas in this file are phrased with data that is bundled into structures like
+`LeftFraction₂` or `LeftFraction₃`. It could have been possible to phrase them
+with "unbundled data". However, this would require introducing 4 or 5 variables instead
+of one. It is also very convenient to use dot notation.
+Many definitions have been made reducible so as to ease rewrites when this API is used.
+
+-/
+
+namespace CategoryTheory
+
+variable {C D : Type*} [Category C] [Category D] (L : C ⥤ D) (W : MorphismProperty C)
+  [L.IsLocalization W]
+
+namespace MorphismProperty
+
+/-- This structure contains the data of two left fractions for
+`W : MorphismProperty C` that have the same "denominator". -/
+structure LeftFraction₂ (X Y : C) where
+  /-- the auxiliary object of left fractions -/
+  {Y' : C}
+  /-- the numerator of the first left fraction -/
+  f : X ⟶ Y'
+  /-- the numerator of the second left fraction -/
+  f' : X ⟶ Y'
+  /-- the denominator of the left fractions -/
+  s : Y ⟶ Y'
+  /-- the condition that the denominator belongs to the given morphism property -/
+  hs : W s
+
+/-- This structure contains the data of three left fractions for
+`W : MorphismProperty C` that have the same "denominator". -/
+structure LeftFraction₃ (X Y : C) where
```

Comment:
Can you extend `LeftFraction₂`? Or is that annoying for some reason?

Reply:
It should be possible, but it would require introducing an alternate constructor `LeftFraction₃.mk'` which reordered arguments. In the tactic state, it may also make annoying expressions appear like `LeftFraction₃.toLeftFraction₂.f`. I am not sure it would be a gain? (`LeftFraction₃` shall be used only in the proof of the associativity of the addition of morphisms in the localization of a preadditive category.)